### PR TITLE
Add RDS proxies

### PR DESCRIPTION
**Pull request by jimleroyer**
# Summary | Résumé

Adding RDS proxies via Terraform. This is part of [the ADR's recommendation](https://github.com/cds-snc/notification-adr/pull/1).

# Test instructions | Instructions pour tester la modification

A similar infrastructure than what is proposed in the Terraform has been setup manually in the *staging* environment. Our goal is to replicate that environment's components with the Terraform code and then target the staging Notify stack to the RDS proxy endpoints.

Also, the RDS proxy have been tested with a local Notify instance running on my computer. The RDS endpoint was tunneled via the jumpbox to my local machine. I had to modify some code of the *notification-api* component to make it work but it was successful. This was done with the RDS manually created component. I will need to replicate the same with the infrastructure automatically created with Terraform once pushed.

# List of changes the Terraform plan will bring in

Compare the [manually created RDS proxy and its artifacts](https://ca-central-1.console.aws.amazon.com/rds/home?region=ca-central-1#proxy:id=notification-canada-ca-staging-rds-proxy-writer) with the list of changes that this Terraform plan will apply once merged:

```
An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create
  ~ update in-place
 <= read (data resources)

Terraform will perform the following actions:

  # aws_rds_cluster.notification-canada-ca will be updated in-place
  ~ resource "aws_rds_cluster" "notification-canada-ca" {
        id                                  = "notification-canada-ca-staging-cluster"
      ~ master_password                     = (sensitive value)
        tags                                = {
            "CostCenter" = "notification-canada-ca-staging"
        }
        # (32 unchanged attributes hidden)
    }

  # aws_secretsmanager_secret.superuser will be created
  + resource "aws_secretsmanager_secret" "superuser" {
      + arn                     = (known after apply)
      + description             = "Database superuser postgres, database connection values"
      + id                      = (known after apply)
      + name                    = "postgres"
      + name_prefix             = (known after apply)
      + policy                  = (known after apply)
      + recovery_window_in_days = 30
      + rotation_enabled        = (known after apply)
      + rotation_lambda_arn     = (known after apply)
      + tags                    = {
          + "CostCenter" = "notification-canada-ca-staging"
        }
      + tags_all                = {
          + "CostCenter" = "notification-canada-ca-staging"
        }

      + rotation_rules {
          + automatically_after_days = (known after apply)
        }
    }

  # aws_secretsmanager_secret_version.superuser will be created
  + resource "aws_secretsmanager_secret_version" "superuser" {
      + arn            = (known after apply)
      + id             = (known after apply)
      + secret_id      = (known after apply)
      + secret_string  = (sensitive value)
      + version_id     = (known after apply)
      + version_stages = (known after apply)
    }

  # module.rds_proxy.data.aws_iam_policy_document.this[0] will be read during apply
  # (config refers to values not yet known)
 <= data "aws_iam_policy_document" "this"  {
      + id   = (known after apply)
      + json = (known after apply)

      + statement {
          + actions   = [
              + "kms:Decrypt",
            ]
          + effect    = "Allow"
          + resources = [
              + "arn:aws:kms:ca-central-1:239043911459:key/a92df413-fc30-4f3e-8047-7433e1a8ad02",
            ]
          + sid       = "DecryptSecrets"

          + condition {
              + test     = "StringEquals"
              + values   = [
                  + "secretsmanager.ca-central-1.amazonaws.com",
                ]
              + variable = "kms:ViaService"
            }
        }
      + statement {
          + actions   = [
              + "secretsmanager:GetRandomPassword",
              + "secretsmanager:ListSecrets",
            ]
          + effect    = "Allow"
          + resources = [
              + "*",
            ]
          + sid       = "ListSecrets"
        }
      + statement {
          + actions   = [
              + "secretsmanager:DescribeSecret",
              + "secretsmanager:GetResourcePolicy",
              + "secretsmanager:GetSecretValue",
              + "secretsmanager:ListSecretVersionIds",
            ]
          + effect    = "Allow"
          + resources = (known after apply)
          + sid       = "GetSecrets"
        }
    }

  # module.rds_proxy.aws_cloudwatch_log_group.this[0] will be created
  + resource "aws_cloudwatch_log_group" "this" {
      + arn               = (known after apply)
      + id                = (known after apply)
      + name              = "/aws/rds/proxy/rds-proxy"
      + retention_in_days = 30
      + tags              = {
          + "CostCenter" = "notification-canada-ca-staging"
        }
      + tags_all          = {
          + "CostCenter" = "notification-canada-ca-staging"
        }
    }

  # module.rds_proxy.aws_db_proxy.this[0] will be created
  + resource "aws_db_proxy" "this" {
      + arn                    = (known after apply)
      + debug_logging          = false
      + endpoint               = (known after apply)
      + engine_family          = "POSTGRESQL"
      + id                     = (known after apply)
      + idle_client_timeout    = 600
      + name                   = "rds-proxy"
      + require_tls            = false
      + role_arn               = (known after apply)
      + tags                   = {
          + "CostCenter" = "notification-canada-ca-staging"
        }
      + tags_all               = {
          + "CostCenter" = "notification-canada-ca-staging"
        }
      + vpc_security_group_ids = [
          + "sg-0e2c3ef6c5c75b74c",
        ]
      + vpc_subnet_ids         = [
          + "subnet-001e585d12cce4d1e",
          + "subnet-08de34a9e1a7458dc",
          + "subnet-0af8b8402f1d605ff",
        ]

      + auth {
          + auth_scheme = "SECRETS"
          + description = "Database superuser postgres, database connection values"
          + iam_auth    = "REQUIRED"
          + secret_arn  = (known after apply)
        }
    }

  # module.rds_proxy.aws_db_proxy_default_target_group.this[0] will be created
  + resource "aws_db_proxy_default_target_group" "this" {
      + arn           = (known after apply)
      + db_proxy_name = "rds-proxy"
      + id            = (known after apply)
      + name          = (known after apply)

      + connection_pool_config {
          + connection_borrow_timeout    = 120
          + max_connections_percent      = 90
          + max_idle_connections_percent = 50
        }
    }

  # module.rds_proxy.aws_db_proxy_endpoint.this["read_only"] will be created
  + resource "aws_db_proxy_endpoint" "this" {
      + arn                    = (known after apply)
      + db_proxy_endpoint_name = "read-only-endpoint"
      + db_proxy_name          = "rds-proxy"
      + endpoint               = (known after apply)
      + id                     = (known after apply)
      + is_default             = (known after apply)
      + tags                   = {
          + "CostCenter" = "notification-canada-ca-staging"
        }
      + tags_all               = {
          + "CostCenter" = "notification-canada-ca-staging"
        }
      + target_role            = "READ_ONLY"
      + vpc_id                 = (known after apply)
      + vpc_security_group_ids = [
          + "sg-0e2c3ef6c5c75b74c",
        ]
      + vpc_subnet_ids         = [
          + "subnet-001e585d12cce4d1e",
          + "subnet-08de34a9e1a7458dc",
          + "subnet-0af8b8402f1d605ff",
        ]
    }

  # module.rds_proxy.aws_db_proxy_endpoint.this["read_write"] will be created
  + resource "aws_db_proxy_endpoint" "this" {
      + arn                    = (known after apply)
      + db_proxy_endpoint_name = "read-write-endpoint"
      + db_proxy_name          = "rds-proxy"
      + endpoint               = (known after apply)
      + id                     = (known after apply)
      + is_default             = (known after apply)
      + tags                   = {
          + "CostCenter" = "notification-canada-ca-staging"
        }
      + tags_all               = {
          + "CostCenter" = "notification-canada-ca-staging"
        }
      + target_role            = "READ_WRITE"
      + vpc_id                 = (known after apply)
      + vpc_security_group_ids = [
          + "sg-0e2c3ef6c5c75b74c",
        ]
      + vpc_subnet_ids         = [
          + "subnet-001e585d12cce4d1e",
          + "subnet-08de34a9e1a7458dc",
          + "subnet-0af8b8402f1d605ff",
        ]
    }

  # module.rds_proxy.aws_db_proxy_target.db_cluster[0] will be created
  + resource "aws_db_proxy_target" "db_cluster" {
      + db_cluster_identifier = "notification-canada-ca-staging-cluster"
      + db_proxy_name         = "rds-proxy"
      + endpoint              = (known after apply)
      + id                    = (known after apply)
      + port                  = (known after apply)
      + rds_resource_id       = (known after apply)
      + target_arn            = (known after apply)
      + target_group_name     = (known after apply)
      + tracked_cluster_id    = (known after apply)
      + type                  = (known after apply)
    }

  # module.rds_proxy.aws_iam_role.this[0] will be created
  + resource "aws_iam_role" "this" {
      + arn                   = (known after apply)
      + assume_role_policy    = jsonencode(
            {
              + Statement = [
                  + {
                      + Action    = "sts:AssumeRole"
                      + Effect    = "Allow"
                      + Principal = {
                          + Service = "rds.amazonaws.com"
                        }
                      + Sid       = "RDSAssume"
                    },
                ]
              + Version   = "2012-10-17"
            }
        )
      + create_date           = (known after apply)
      + force_detach_policies = true
      + id                    = (known after apply)
      + managed_policy_arns   = (known after apply)
      + max_session_duration  = 43200
      + name                  = "rds-proxy-to-secrets-role"
      + path                  = "/"
      + tags                  = {
          + "CostCenter" = "notification-canada-ca-staging"
        }
      + tags_all              = {
          + "CostCenter" = "notification-canada-ca-staging"
        }
      + unique_id             = (known after apply)

      + inline_policy {
          + name   = (known after apply)
          + policy = (known after apply)
        }
    }

  # module.rds_proxy.aws_iam_role_policy.this[0] will be created
  + resource "aws_iam_role_policy" "this" {
      + id     = (known after apply)
      + name   = "rds-proxy-to-secrets-policy"
      + policy = (known after apply)
      + role   = (known after apply)
    }
```

# Help requested | Aide requise

Help from the SEQ folks for review would be tremendously helpful.

# Reviewer checklist | Liste de vérification du réviseur

This is a suggested checklist of questions reviewers might ask during their
review | Voici une suggestion de liste de vérification comprenant des questions
que les réviseurs pourraient poser pendant leur examen :

- [ ] Does this meet a user need? | Est-ce que ça répond à un besoin utilisateur?
- [ ] Is it accessible? | Est-ce que c’est accessible?
- [ ] Is it translated between both offical languages? | Est-ce dans les deux
      langues officielles?
- [ ] Is the code maintainable? | Est-ce que le code peut être maintenu?
- [ ] Have you tested it? | L’avez-vous testé?
- [ ] Are there automated tests? | Y a-t-il des tests automatisés?
- [ ] Does this cause automated test coverage to drop? | Est-ce que ça entraîne
      une baisse de la quantité de code couvert par les tests automatisés?
- [ ] Does this break existing functionality? | Est-ce que ça brise une
      fonctionnalité existante?
- [ ] Should this be split into smaller PRs to decrease change risk? | Est-ce
      que ça devrait être divisé en de plus petites demandes de tirage (« pull
      requests ») afin de réduire le risque lié aux modifications?
- [ ] Does this change the privacy policy? | Est-ce que ça entraîne une
      modification de la politique de confidentialité?
- [ ] Does this introduce any security concerns? | Est-ce que ça introduit des
      préoccupations liées à la sécurité?
- [ ] Does this significantly alter performance? | Est-ce que ça modifie de
      façon importante la performance?
- [ ] What is the risk level of using added dependencies? | Quel est le degré de
      risque d’utiliser des dépendances ajoutées?
- [ ] Should any documentation be updated as a result of this? (i.e. README
      setup, etc.) | Faudra-t-il mettre à jour la documentation à la suite de ce
      changement (fichier README, etc.)?

### CODE REVIEW

File: aws/rds/variables.tf [hunk 1]
```diff
@@ -2,6 +2,14 @@ variable "eks_cluster_securitygroup" {
...
+  type = string
+}
+
+variable "rds_server_db_user" {
+  type = string
+}
+
 variable "rds_cluster_password" {
   type = string
 }
```
Comment (CalvinRodo) [inline aws/rds/variables.tf:15]: this should be marked as sensitive so it doesn't accidentally get output in some logs. 

Adding `sensitive = true` should do this. 
https://learn.hashicorp.com/tutorials/terraform/sensitive-variables

**Review by CalvinRodo — CHANGES_REQUESTED** [review]
Just see the comment about the sensitive string. 

I'm always a little hesitant about using 3rd party modules just make sure you review the infrastructure that is created by this PR.

File: aws/rds/variables.tf (hunk 1 above)
Comment (jimleroyer) [reply aws/rds/variables.tf]: Nice catch, just pushed the change.

**Review by CalvinRodo — APPROVED** [review]
:taco::cat:

**Pull request merged**